yt/frontends/flash/fields.py

```diff
@@ -148,44 +150,67 @@ def _eint(field, data):
+
         ## Derived FLASH Fields
+
         def _nele(field, data):
             Na_code = data.ds.quan(Na, '1/code_mass')
             return data["flash","dens"]*data["flash","ye"]*Na_code
+
         self.add_field(('flash','nele'),
                        sampling_type="cell",
                        function=_nele,
                        units="code_length**-3")
+
         self.add_field(('flash','edens'),
                        sampling_type="cell",
                        function=_nele,
                        units="code_length**-3")
+
         def _nion(field, data):
             Na_code = data.ds.quan(Na, '1/code_mass')
             return data["flash","dens"]*data["flash","sumy"]*Na_code
+
         self.add_field(('flash','nion'),
                        sampling_type="cell",
                        function=_nion,
                        units="code_length**-3")
 
         if ("flash", "abar") in self.field_list:
-            self.add_output_field(("flash", "abar"),
-                                  sampling_type="cell",
-                                  units="1")
-        else:
+            self.alias(("gas", "mean_molecular_weight"), ("flash", "abar"))
+        elif ("flash", "sumy") in self.field_list:
             def _abar(field, data):
                 return 1.0 / data["flash","sumy"]
-            self.add_field(("flash","abar"),
+            self.add_field(("gas", "mean_molecular_weight"),
+                           sampling_type="cell",
+                           function=_abar,
+                           units="")
+        elif "eos_singlespeciesa" in self.ds.parameters:
+            def _abar(field, data):
+                return data.ds.parameters["eos_singlespeciesa"]*data["index", "ones"]
+            self.add_field(("gas", "mean_molecular_weight"),
                            sampling_type="cell",
                            function=_abar,
-                           units="1")
+                           units="")
 
-        def _number_density(fields,data):
-            return (data["nele"]+data["nion"])
-        self.add_field(("gas","number_density"),
-                       sampling_type="cell",
-                       function=_number_density,
-                       units=unit_system["number_density"])
+        if ("flash", "nele") in self.field_list and ("flash", "nion") in self.field_list:
+            def _number_density(field, data):
+                return data["flash", "nele"]+data["flash","nion"]
+
+            self.add_field(("gas", "number_density"),
+                           sampling_type="cell",
+                           function=_number_density,
+                           units=unit_system["number_density"])
+
+        else:
+            def _number_density(field, data):
+                Na_code = data.ds.quan(Na, '1/code_mass')
+                return data["flash", "dens"]*Na_code/data["gas", "mean_molecular_weight"]
+
+            self.add_field(("gas", "number_density"),
```

Comment:
Minor, but you could indent this `self.add_field` call one to the left and remove the identical call inside the if statement above.